server/src/main/java/org/eclipse/lsp/cobol/service/CopybookNameServiceImpl.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Broadcom, Inc. - initial API and implementation
+ *
+ */
+package org.eclipse.lsp.cobol.service;
+
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+import com.google.inject.Singleton;
+import lombok.extern.slf4j.Slf4j;
+import org.eclipse.lsp.cobol.jrpc.CobolLanguageClient;
+import org.eclipse.lsp.cobol.service.utils.FileSystemService;
+import org.eclipse.lsp4j.WorkspaceFolder;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+import static org.eclipse.lsp.cobol.service.utils.SettingsParametersEnum.CPY_LOCAL_PATHS;
+
+/**
+ * This service processes all the copybook names present in the local directory. The service also
+ * caches copybook to reduce filesystem load.
+ */
+@Slf4j
+@Singleton
+public class CopybookNameServiceImpl implements CopybookNameService {
+  private final FileSystemService files;
+  private final Provider<CobolLanguageClient> clientProvider;
+  private final SettingsService settingsService;
+  private List<String> resolvedCopybookNames;
+
+  @Inject
+  public CopybookNameServiceImpl(
+      SettingsService settingsService,
+      FileSystemService files,
+      Provider<CobolLanguageClient> clientProvider) {
+    this.settingsService = settingsService;
+    this.files = files;
+    this.clientProvider = clientProvider;
+    this.resolvedCopybookNames = new ArrayList<>();
```

Comment:
Or you can do an atomic link assignment for `resolvedCopybookNames`. So instead of: `.forEach(resolvedCopybookNames::add);` you can collect names into a list and assign the `resolvedCopybookNames` to it. And in that case you don't need to synchronize any methods in this class.

Reply:
Makes sense. If the list will be immutable, then we don't even need to copy the list on returning.